feat: pooler docs (#100)
neondatabase/website: content/docs/get-started-with-neon/connection-pooling.md

---
title: Connection pooling
---

## Postgres connection limits

> Each postgres connection creates a new process in the operating system, which occuppies some amount of resources -- memory, open file descriptors, etc. Hence postgres has limits on amount of open connections. I in our free tier we allow 20 simultaneous postgres connections.

Reviewer: ```suggestion
Each postgres connection creates a new process in the operating system, which occupies some resources -- memory, open file descriptors, etc. Hence Postgres has limits on the amount of available connections. In our Free Tier we allow 20 simultaneous postgres connections.
```

> Some apps tends to open a lot more connections and keep most of them inactive. This usually happens either because of database driver limitation or due to the lots of deployed instances of the app. To deal with such situatiations we use connection pooler.

Reviewer: ```suggestion
Some apps tend to open a lot more connections and keep most of them inactive. This usually happens either because of database driver limitation or due to the lots of deployed instances of the app. To deal with such situatiations, we use connection pooler.
```

## Pooling in Neon

> Connection pooler is deploed near the postgres and allows to accept up to 500 connections. This connection then routed to the smaller amount of real postgres connections. We use `pgbouncer` in `transaction mode` for connection pooling.

Reviewer: ```suggestion
Connection pooler is deployed near the Postgres and allows to accept up to 500 connections. These connections are then routed to the smaller amount of real postgres connections. We use `pgbouncer` in `transaction mode` for connection pooling.
```

> To enable pooling go to the project settings and tick `Enable pooling` checkbox under a `General` section.

Reviewer: ```suggestion
To enable pooling go to the project settings and tick the `Enable pooling` checkbox under the `General` section.
```

### Limitations

> Some of the database features like prepared statements and `LISTEN/NOTIFY` functionality won't work with connection pooling. To get a full list of limitations see `Transaction pooling` column in features table at <https://www.pgbouncer.org/features.html>.

Reviewer: ```suggestion
Some database features like prepared statements and `LISTEN/NOTIFY` functionality won't work with connection pooling. To get a complete list of limitations, see the `Transaction pooling` column in the features table at <https://www.pgbouncer.org/features.html>.
```
